mcs/class/System/System.Net.Sockets/Socket.cs

```diff
@@ -2681,14 +2681,12 @@ SocketAsyncResult ValidateEndIAsyncResult (IAsyncResult ares, string methodName,
 
 		void QueueIOSelectorJob (SemaphoreSlim sem, IntPtr handle, IOSelectorJob job)
 		{
-			sem.WaitAsync ().ContinueWith (t => {
-				if (CleanedUp) {
-					job.MarkDisposed ();
-					return;
-				}
-
-				IOSelector.Add (handle, job);
-			});
+			sem.Wait();
```

Comment:
Thanks. I made a mistake!
What about providing a fast path without allocating `Task<Action>` ?
```
var task = sem.WaitAsync();
if ( task.IsCompleted) {
        if (Cleanup) 
       // ...
}
else 
{
        task.ContinueWith( t => {
                  if (Cleanup) 
                  // ...
        }
}
```

The idea behind this change is, reading TCP socket in single thread is pretty common, like
```
BeginReceive(..., Callback);
Callback = ar => {
        BeginReceive(..., Callback)

}
```
No multi threads working on reading/writing single socket, no lock contention, therefore the `task.IsCompleted` fast path always stands.